### ISSUE TITLE
Shakespeare's style writing generation model cannot be exported with TensorFlow 2.3

### ISSUE THREAD

**Issue by nobutoba** [tensorflow/tensorflow#46867]
[text_generation.ipynb]: https://github.com/tensorflow/docs/blob/master/site/en/tutorials/text/text_generation.ipynb
[Colab notebook]: https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/text/text_generation.ipynb
## URL(s) with the issue:

- TensorFlow Tutorial: Text generation with an RNN (the one that produces Shakespeare's style writing using RNN).

    https://github.com/tensorflow/docs/blob/master/site/en/tutorials/text/text_generation.ipynb

- Accompanying Colab.

    https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/text/text_generation.ipynb

## Description of issue (what needs changing):

With TensorFlow 2.4, the notebook [text_generation.ipynb] above works as expected (without throwing any error), but this is not the case with TensorFlow 2.3.

A trivial solution is to upgrade to TensorFlow 2.4. However, this would require, when using pre-built TensorFlow, switching to CUDA 11 (cf. https://www.tensorflow.org/install/source#gpu). In a cloud environment I use from time to time, CUDA 11 is not yet supported; this is where my trouble comes from.

So the notebook better works well with TensorFlow 2.3 or higher.

### Clear description

To reproduce the issue:

1. Open the [Colab notebook].
1. Select Runtime > Change runtime type > GPU.
1. insert the line `!pip install 'tensorflow<2.4'` before the line `import tensorflow as tf`.
1. Select Runtime > Run all.

Environmental setup, text preprocessing, model building, and training run smoothly, but after the following cell in the "Export the generator" section, a `ValueError` is raised:

```
tf.saved_model.save(one_step_model, 'one_step')
one_step_reloaded = tf.saved_model.load('one_step')
```

```
WARNING:tensorflow:Skipping full serialization of Keras layer <__main__.OneStep object at 0x7f020a19e860>, because it is not built.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Layer.updates (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
INFO:tensorflow:Assets written to: one_step/assets
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-45-5089d08e324f> in <module>()
      1 tf.saved_model.save(one_step_model, 'one_step')
----> 2 one_step_reloaded = tf.saved_model.load('one_step')

9 frames
/usr/local/lib/python3.6/dist-packages/tensorflow/python/training/saving/saveable_object_util.py in _add_saveable(saveables, seen_ops, saveable)
    329   if saveable.op in seen_ops:
    330     raise ValueError("The same saveable will be restored with two names: %s" %
--> 331                      saveable.name)
    332   saveables.append(saveable)
    333   seen_ops.add(saveable.op)

ValueError: The same saveable will be restored with two names: ids_from_chars/_table/.ATTRIBUTES/table
```

Just so you know, this notebook is not compatible with TensorFlow 2.2 for the very reason that `StringLookup` is not supported; running it with TensorFlow 2.2 raises an exception as follows:

```
ids_from_chars = preprocessing.StringLookup(
    vocabulary=list(vocab))
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-aea607cfc6f7> in <module>()
----> 1 ids_from_chars = preprocessing.StringLookup(
      2     vocabulary=list(vocab))

AttributeError: module 'tensorflow.keras.layers.experimental.preprocessing' has no attribute 'StringLookup'
```

So the `ValueError` in concern is an exception peculiar to TensorFlow 2.3. So far I have no idea why the notebook is not compatible with TensorFlow 2.3, and any help or a single hint will be very much appreciated.

**Comment by amahendrakar** [comment]
@nobutoba,
As you've mentioned, I was able to reproduce the error with TF v2.3 and looks like this error was fixed in TensorFlow v2.4.

A hack that worked for me was to edit the module throwing the error i.e. changing line number 329 in the file
`/usr/local/lib/python3.6/dist-packages/tensorflow/python/training/saving/saveable_object_util.py` from

`if saveable.op in seen_ops:`
to 
`if saveable.op is not None and saveable.op in seen_ops:`

After the edit, save the changes, restart the runtime and check if it works. 

Please be advised that this is just a workaround because of the restrictions on your machine and may result in unexpected behavior.

References: 
- [Link 1](https://colab.research.google.com/gist/amahendrakar/054e49902bddd0182603cbbeabefb731/46867.ipynb#scrollTo=3Grk32H_CzsC&line=2&uniqifier=1) - Colab notebook without the edit
- [Link 2](https://colab.research.google.com/gist/amahendrakar/225045bbc9ebd71c4a8aee62a8302d9c/46867-withedit.ipynb#scrollTo=jE8Pdbpoiw0d&line=1&uniqifier=1) - Colab notebook with edit 
- https://github.com/tensorflow/tensorflow/commit/2d9fb19f8d0672245c2becd06b05a59c1d43c33e - Commit ID

Hope this helps. Thanks :)

**Comment by nobutoba** [comment]
@amahendrakar Thank you very much, your explanation was very thorough 👍
